### ISSUE TITLE
Residual f32 / route-mismatch coordinate constructions outside the binary rasterization class

### ISSUE THREAD

**Issue by bk-squared** [bk-squared/rfx#833]
Two residuals the adversarial review of the #802/#807 fix found and deliberately left out of that PR's scope (both benign for realized masks; neither flips a cell):

1. **`rfx/geometry/smoothing.py::_yee_coords`** still builds sample coordinates as `(jnp.arange(n) - pad) * dx` in JAX's default dtype, so the OPT-IN Kottke subpixel-smoothing SDF path samples at f32, flag-dependent positions. No binary plane flip is possible (the fill fraction is continuous, and its staircase fallback uses the fixed `shape.mask`), but it is the same formula the #802 fix removed everywhere else — the smoothed ε near a boundary differs at ~1e-7 relative between x64 flags. Route it through the shared exact node builder for consistency.

2. **`rfx/fidelity.py::_node_arrays`** builds report-side node arrays by cumsum (f32 stores on NU, f64 summation on uniform) instead of the exact spine, leaving ~1-ulp `face_residual` noise in fidelity reports (the ~1e-12 µm leaves visible in the re-captured snapshot). Reporting-only; realized masks use the fixed producers.

Neither affects solve results. Fix both by routing through the exact host-f64 node builder introduced by the #802 fix; acceptance = smoothed-ε x64-invariance on a boundary voxel, and zero face_residual on exactly-on-node faces in the fidelity report.

**Comment by bk-squared** [comment]
Item 2 (fidelity node-array cumsum noise) closed by #845 (squash 1c38b0d7): the report now reads the rasterizer's exact host-float64 node line, and sub-1e-9-cell face residuals are reported as exactly 0.0 (fe2fe8f9 — the last float64 digit of k*dz - face was host-dependent and reached the snapshot). Item 1 (rfx/geometry/smoothing.py::_yee_coords float32 sampling on the opt-in Kottke path) is still open.

**Comment by bk-squared** [comment]
## Narrowed on `9a187818`: item 2 is fixed, item 1 is the live residue with a measured number

**Item 2 fixed** — `rfx/fidelity.py:65` names "#833 item 2, measured 2026-09-02"; the repro now prints `face residuals (0.0, 0.0) um`.

**Item 1 still open**, verified at the line. `rfx/geometry/smoothing.py:233` is unchanged:

```python
x = (jnp.arange(nx) - pad_x) * dx
```

`jnp.arange` returns `float32` under `JAX_ENABLE_X64=0`, so the node line is built in single precision on that lane. Measured max deviation from the exact node coordinates: **7.45e-10 m under x64=0, exactly 0.0 under x64=1** — i.e. the construction is still route-dependent, which is the class #802/#807 closed elsewhere by moving to exact host-f64 node coordinates (`rfx/geometry/rasterize_grid.py:69-80`).

**Re-scoped ask**: item 1 only, and the fix is the pattern already committed in `rasterize_grid.py` — compute the node line in host f64 and hand it in, rather than deriving it from `jnp.arange` inside the traced function. Worth checking in the same pass whether any other smoothing-path coordinate is built the same way.

Cross-reference: this is also the unsettled decision 8 on #803.

**Comment by bk-squared** [comment]
사용자 요청으로 2026-09-13 백로그 정리의 종료 판단을 철회하고, 재검토를 위해 다시 엽니다.

재확인할 항목: 병합된 좌표 경로 수정과 x64 설정별 좌표·정밀도·AD 검증이 원래 요구를 충족하는지 종료 근거를 재확인합니다. 기존 수정이 잘못됐다고 확정하는 댓글은 아닙니다.

기존 수정·측정·반례 기록과 담당 agent/PR의 작업 범위를 유지합니다. Codex의 추가 코드 수정·실험·PR 작업은 중단했으며, 이번 처리는 댓글과 이슈 상태 복구에 한정합니다.

**Comment by bk-squared** [comment]
🔒 **Claimed** (session 452c7a2e / Fable, 2026-09-16, PI 지시로 비어 있는 재검토 이슈 착수). 9/13 재오픈 시 요청된 '재확인 항목'을 독립 재검증하고, 종료 근거가 서면 근거와 함께 닫거나 남은 작업을 구체화해 이어서 처리합니다. 다른 세션은 시작 전에 여기에 코멘트 부탁합니다.

**Comment by bk-squared** [comment]
## Independent re-verification of the 2026-09-13 reopen items (session 452c7a2e, 2026-09-16)

Measured on main `6d721a56` (HEAD) and on `3f3ed7e0` (parent of ab280a38 / PR #998), CPU host, local `.venv` (jax 0.10.2), `rfx.__file__` printed per process. Auditor and an adversarial refuter measured independently; the numbers below reproduced bit-for-bit between them.

### Item 2 (fidelity node arrays) — settled

`rfx/fidelity.py::_node_arrays` reads `_uniform_axis_nodes` / `coords_from_nonuniform_grid` (the exact host-f64 producers). Fixture `Simulation(freq_max=6e9, domain=(10,10,15) mm, dx=0.5 mm, cpml_layers=8, all-CPML)`, NU lane with `dz_profile=np.full(30, D)`; no warnings emitted.

| check | uniform lane | NU lane |
|---|---|---|
| max \|report node − rasterizer node\| (x, y, z) | 0.0 m | 0.0 m |
| `face_residual_um` for on-node faces `10D..11D`, `5.0e-3..5.5e-3`, `(4D+6D)..11D`, reported | (0.0, 0.0) | (0.0, 0.0) |
| same, RAW with `_zero_tie` snap disabled | (0.0, 0.0) | (0.0, 0.0) |
| off-node sanity `hi = 11.5D` | (0.0, 250.00000000000023) um, n_cells 200 | same |

Both x64 flags. The zero residual is raw, not the snap. Closed by #845 as recorded.

### Item 1 (uniform Kottke coordinates) — NOT at the stated acceptance

Fixture: `Grid(freq_max=1e9, domain=(7,5,3) mm, dx=2.54e-4, cpml_layers=2)` → shape (33,25,17), pads (2,2,2), **no Grid warnings**; `Box lo=(1.3,1.1,0.7) mm hi=(4.7,3.9,2.3) mm`, ε 4.0 (faces off-node so every interface voxel has a partial fill).

Coordinates (`_yee_coords`, max |sample − exact node|):

| tree | x64=0 | x64=1 |
|---|---|---|
| 3f3ed7e0 (pre-fix, `jnp.arange` formula) | 5.318e-10 / 4.974e-10 / 2.487e-10 m (x/y/z), not the f32 rounding of the spine | 0.0 m |
| 6d721a56 (ab280a38) | 2.196e-10 / 2.117e-10 / 1.058e-10 m, **exactly the f32 rounding of the spine** | 0.0 m |

Smoothed ε, `eps(x64=0)` vs `eps(x64=1)` on all voxels (every 576 / 603 / 616 interface voxel differs in both trees):

| array | pre-fix max rel (f32 ulps) | HEAD max rel (f32 ulps) |
|---|---|---|
| ex | 1.019e-06 (14) | 1.131e-06 (15) |
| ey | 5.661e-06 (48) | 6.467e-07 (5) |
| ez | 5.661e-06 (48) | 1.025e-06 (11) |
| inv-ε ix | 1.054e-06 (11) | 1.150e-06 (12) |
| inv-ε iy | 5.669e-06 (94) | 6.582e-07 (11) |
| inv-ε iz | 5.669e-06 (94) | 9.651e-07 (13) |

Sample voxel (7,6,5): `eps_x = 1.5078747272491455` at x64=0 (both trees) vs `1.5078740157480277` at x64=1. The acceptance line on this issue is "smoothed-ε x64-invariance on a boundary voxel"; it is not met.

Decomposition (x64=1 f64 arithmetic with only the coordinates substituted): f32-rounded exact nodes (= HEAD's x64=0 coordinates) give max rel ex 5.275e-07, ey 4.634e-07, ez 5.727e-07; the old f32 `arange` coordinates give 1.683e-06 / 4.972e-06 / 4.972e-06. Coordinate rounding explains roughly half of HEAD's flag dependence; the other half is float32 SDF / Kottke arithmetic. Coordinate routing alone cannot reach invariance — the floor for a coordinate-only fix is the f32 arithmetic of the path, and only a host-f64 fill/normal computation for concrete shapes (bound ≈ 0.5 f32 ulp ≈ 6e-8 rel) can.

Invariance witness PR #998 skipped: at **x64=0** ab280a38 changed the forward operator on the opt-in uniform smoothing lane (HEAD vs 3f3ed7e0): ex max|Δ| 3.457e-06 on 478 voxels, ey 5.484e-06 / 434, ez 5.484e-06 / 382; inv-ε ix 1.371e-06 / 478, iy 5.364e-06 / 434, iz 5.364e-06 / 382. These arrays feed `rfx/runners/uniform.py` and `rfx/sparams/waveguide.py` when `subpixel_smoothing=True`. At x64=1 all six arrays are bit-identical (0 voxels differ). The PR body's "changes no solver operator" holds only at x64=1; the CHANGELOG line ("positions varied with the active JAX precision setting" removed) overstated the result. Both are corrected in the PR below.

Third `_yee_coords` consumer the PR did not list: `compute_conformal_weights_sdf` (`rfx/geometry/conformal.py:96`, conformal-PEC lane, weights feed Cb). Measured HEAD vs pre-fix, both flags, sphere PEC (54 partial cells) and box PEC: `w_ex/w_ey/w_ez` max|Δ| 0.0, 0 voxels differ. A future change to `_yee_coords` touches that lane too.

AD re-confirmation: `jax.grad` of Σ eps_x through `compute_smoothed_eps` w.r.t. a Sphere radius is finite and FD-validated — HEAD x64=1 grad 2272071.1593872565 vs central FD (h=1e-8) 2272071.16304271 (rel 1.6e-9); x64=0 grad 2272071.75 (cross-flag 2.6e-7 rel; pre-fix x64=0 2272072.25). The same gradient w.r.t. a **Box** shift is NaN on both trees and both flags while the central FD is finite (−700273.02 /m at x64=1; −700683.56 HEAD x64=0; −699462.88 pre-fix x64=0). Cause: `jnp.sqrt(ox²+oy²+oz²)` at 0 after the `max(·, 0)` clamp in `_sdf_box` (same form in `_sdf_cylinder`); pre-existing, not touched by ab280a38, and the production design-AD path `kottke_inv_eps_from_occupancy` never calls `_yee_coords`. Filed separately as #1085.

### Other smoothing-path coordinate built the same way (the narrowing comment's ask) — found and fixed

`compute_smoothed_eps_nonuniform` built its cell centres as `node + f32(store)/2` from the float32 solver store. On the graded fixture (`dz = 20×D | 10×D/2 | 20×D`, D = 0.3048 mm, `cpml_layers=8`, shape 47×37×67; build warning verbatim: `dz_profile has max adjacent cell ratio 2.000 (> 1.4, the validated multi-band grading cap — docs/guides/support_matrix.md, 'Multi-band graded mesh'). This may cause numerical reflections. Use rfx.smooth_grading(dz_profile) to fix.`) the centres sat **3.704e-12 m** off `node + d_exact/2` at x64=1 on all axes, 5.18e-10 / 4.44e-10 / 1.10e-09 m at x64=0 (not the f32 rounding either). PR #1088 derives them from the float64 spine in host f64, cast once: **0.0 m** at x64=1, exactly `f32(exact)` at x64=0. Forward-operator delta on that lane (after vs before): ez max|Δ| 3.576e-06 on 171 voxels at x64=0, 1.354e-08 (7.5e-09 rel) on 342 voxels at x64=1; ex/ey 0 for a Box on nodes; an off-node Box additionally moves ex by 7.868e-06 / 63 voxels (x64=0), 6.041e-08 / 126 (x64=1). No committed NU test or golden moved; the reduces-to-uniform gate (1e-5) is untouched (worst gap now 1.431e-06 at x64=0, 1.232e-07 at x64=1).

### What PR #1088 does

1. NU centres from the float64 spine (above), with a contract test at both flags.
2. A measured **drift pin** for the uniform lane's cross-flag ε (gate 1.7e-06 / 1.0e-06 / 1.6e-06 rel = envelope × 1.5 rounded up; docstring says it records the state after ab280a38 and is not the acceptance).
3. CHANGELOG: new entry, and a one-sentence correction of the ab280a38 entry.

It does not implement invariance.

Test state reported with it: `tests/unit/geometry/` + the two subpixel boundary files + the NU-referencing `tests/unit/runners/` files — x64=0 `425 passed, 7 skipped` rc=0; x64=1 `5 failed, 420 passed` — the five (`test_compute_inv_eps_tensor_diag_returns_float32`, four `test_distributed_nu_composition.py` scan-carry dtype failures) fail identically on a clean archive of main `6d721a56` at x64=1, so they are pre-existing x64=1-lane red, not moved by the PR, and outside its scope.

### Decision for the PI

- **(a)** Compute Kottke fill fractions and interface normals in host float64 for concrete shapes (Box/Sphere/Cylinder), casting to the JAX dtype only at the end, so that `eps(x64=0) == f32(eps(x64=1))` bitwise (or ≤ 0.5 f32 ulp) on every interface voxel for `compute_smoothed_eps` and `compute_inv_eps_tensor_diag`; then tighten the drift pin to that bound. Estimate 0.5–1 day; the traced-shape path must keep JAX arithmetic, so the split has to be explicit.
- **(b)** Ratify the weaker acceptance "coordinates consume the shared spine; ε agrees to the f32 lane's arithmetic floor", with the measured envelope above quoted on this issue, and close #833 on PR #1088. Estimate 1 hour (comment + closing).

Until one is chosen, #833 stays open with item 1 in this measured state.

**Comment by bk-squared** [comment]
**PI decision (2026-09-16): option (a)** — host-float64 fill fractions and normals for concrete shapes. Implemented in #1095 (branch `fix/833-host-f64-fill-fractions`).

**What the PR does.** When every input to `compute_smoothed_eps` / `compute_inv_eps_tensor_diag` / `compute_smoothed_eps_nonuniform` is concrete (host node spine, Python-float shape parameters and `eps_r`), the SDF → fill-fraction → analytic-normal → Kottke chain runs in numpy float64 and is cast to the active JAX dtype once at the end. The SDF/normal functions take an array backend `xp`; one Kottke body serves all three entry points; a traced input (Sphere radius under `jax.grad`, traced mesh axis) takes the same body through `jax.numpy` as before. Formulas, output dtypes and shapes unchanged. NU fill-fraction normalisation reads the float64 spine instead of the float32 store.

**Measured invariance (acceptance).** `eps(x64=0) == float32(eps(x64=1))` bitwise on every voxel: Box / Sphere / Cylinder-z / Cylinder-x / mixed-ε group on the boundary-voxel fixture (dx 2.54e-4, 14 025 voxels); `compute_inv_eps_tensor_diag` with the dielectric Box alone and with PEC Sphere + PEC Cylinder; staircase-fallback and fallback+Sphere; Box / Sphere / Cylinder / mixed on the graded NU fixture (47×37×67). Before (b8788646): 15/5/11 ulps on the Box (511/603/538 voxels), Sphere up to 80 ulps, and on the NU Box **12 voxels differed by |Δε| 0.99** — voxels exactly equidistant from the x and z faces where float32 broke the nearest-face tie differently per flag (ε 2.105 vs 3.100). The drift pin in `test_smoothing_coordinate_contract.py` is replaced by an 8-case bitwise lock (`assert_array_equal`) plus a traced-radius fallback test.

**x64=1 is reproduced.** Uniform lane bit-identical or ≤ 4 f64 ulps (numpy vs XLA reassociation); NU lane ≤ 2.2e-7 rel on interface voxels because the geometric-mean cell length is now formed in float64 from the spine (the old float32 value was 2.6e-7 rel off on every cell). End-to-end 100-step CPML run with an off-node ε-4 Box: all fields bit-identical at x64=1 for `True` / `"kottke_pec"` / `False`.

**Operator delta at x64=0 (the default precision, opt-in lane).** The smoothed ε is now the float32 image of the float64 result: Box ex/ey/ez max|Δ| 1.9e-06 / 1.2e-06 / 1.3e-06 (15 / 5 / 11 ulps), inverse-ε 12 / 11 / 13 ulps, Sphere up to 1.9e-05 (80 ulps), Cylinder up to 53 ulps; NU Box 16 ulps plus the 12 tie voxels above. End-to-end fields move ≤ 8.5e-7 of the field maximum; `subpixel_smoothing=False`, the staircase fallback, `compute_conformal_weights_sdf` and the traced path are bit-identical at both flags.

**Tests.** `tests/unit/geometry/` 311 passed at x64=0; at x64=1 exactly the five pre-existing reds (`test_compute_inv_eps_tensor_diag_returns_float32` + four `test_distributed_nu_composition.py`), 323 passed, nothing new; 363 passed across the boundary / sparams / runner / contract lanes that touch smoothing. No committed gate moved.

**Left, recorded not fixed.** The PEC-limit branch of `compute_inv_eps_tensor_diag` still promotes to float64 at x64=1 (pre-existing red, dtype contract reproduced not changed). `compute_conformal_weights_sdf` has its own cross-flag drift (w_ez rel 3.7e-4 at a near-zero weight) — conformal-PEC lane, outside this issue's acceptance, no issue filed yet. With the lock green for every function and SDF family, #1095 carries `Fixes #833`.

**Comment by bk-squared** [comment]
Correction to the tie description in my comment above (review finding on #1095): the 12 NU Box voxels are 0.2 D inside both the **x-lo** and **z-hi** faces (not z-lo), `rx == rz == −0.2 D` in exact arithmetic. `_normal_box` picks x on `rx >= rz`, which the exact tie would satisfy (ε 2.105); in float64 the two distances round differently, `rz − rx` = 64 f64 ulps, so z wins (ε 3.100). That is a deterministic rounding result, not a convention; base x64=1 already gave 3.100, and only the float32 evaluation (x64=0 before this PR, and the traced `jax.numpy` path at x64=0 still) gives 2.105.

Two more qualifications from the review, now in the PR body and CHANGELOG: "x64=1 reproduced" holds for float32-representable ε; the old accumulators were float32 even at x64=1, so `background_eps=2.2` moves every non-interface voxel by 2.167e-08 rel (2.2 vs float32(2.2)). Reviewer-measured build cost on a 165³ grid: peak RSS delta 943 MB (this branch) vs 1327 MB (base) at x64=0, 943 vs 837 MB at x64=1, about 4× faster. A traced mask-only shape (regressed in the first commit) is routed back through `jax.numpy` and pinned at both flags.

**Comment by bk-squared** [comment]
Recorded limitation (PR #1095 review, item 7): on exact Kottke-normal ties (`rx == rz`, the 12 NU Box voxels above) the concrete host-float64 path now returns ε 3.100 at both flags, while the traced `jax.numpy` path at x64=0 still returns 2.105 — concrete vs traced disagree by 0.99 on such voxels. Before #1095 the two x64=0 paths agreed (both 2.105). Making the tie rule explicit in `_normal_box` would move x64=1 values on tie voxels, so it is a separate formula change and is deliberately not part of #1095. No new issue is filed; this comment is the record.
